### ISSUE TITLE
Support for clusters with OwnerReferencesPermissionEnforcement admission plugin enabled

### ISSUE THREAD

**Issue by erikgb** [clastix/capsule#773]
# Bug description

We are performing a Capsule POC on OpenShift, and it turns out that Capsule misses RBAC configuration to support clusters with [OwnerReferencesPermissionEnforcement admission plugin](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#ownerreferencespermissionenforcement) enabled. OpenShift enables this by default.

# How to reproduce

Pre-requisite: A Kubernetes cluster with OwnerReferencesPermissionEnforcement admission plugin enabled.

1. Establish a new tenant
2. As a tenant owner, attempt to create a new namespace
3. The create namespace request is rejected with the following error message: `namespaces "erikbo-egb-test" is forbidden: cannot set blockOwnerDeletion if an ownerReference refers to a resource you can't set finalizers on`

# Expected behavior

The tenant owner is allowed to create a new namespace without any errors.

I am pretty sure that the error is caused by the Capsule namespace admission webhook. It adds an owner reference to the namespace, pointing to the tenant resource with `blockOwnerDeletion: true`. Which a `capsule-namespace-provisioner` is not allowed to do. I was able to work around the issue by adding a new cluster role with the missing permissions and bind this new cluster role to the `capsuleUserGroups`.

````yaml
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: capsule-namespace-provisioner-fix
rules:
  - verbs:
      - update
    apiGroups:
      - capsule.clastix.io
    resources:
      - tenants/finalizers
````

# Logs

If applicable, please provide logs of `capsule`.

I don't think the logs are relevant in this case.

# Additional context

- Capsule version: latest + @prometherion Flux impersonate hot-fix
- Helm Chart version: N/A
- Kubernetes version: Openshift 4.12.13 (K8s v1.25.8+27e744f)

**Comment by oliverbaehler** [comment]
Hi @erikgb, thanks for the issue. I will create a dev version so you can continue with your poc (should have time later today).

May I ask to which version you are refering to with "latest + @prometherion Flux impersonate hot-fix" ? 

Are we talking about this: https://github.com/clastix/capsule/pull/768?

**Comment by erikgb** [comment]
Hi @oliverbaehler, thanks for responding. I can understand why you ask about the version. 😆 I just realize that this is not really relevant to this issue. I will explain: We are installing Capsule and Capsule Proxy with the latest version of the Helm charts. But because of issues passing through capsule-proxy from a Flux GitOps reconcile, @prometherion built a custom latest version of capsule-proxy after the community meeting on Tuesday. See https://kubernetes.slack.com/archives/C03GETTJQRL/p1686665245918379?thread_ts=1686311373.987009&cid=C03GETTJQRL for exact details.

But we run a released version of capsule: 0.3.2

**Comment by oliverbaehler** [comment]
@erikgb Would you be willing to test it with our fix? Its a bit tricky for me because I don't have an OpenShift laying around.. :D 

```
git clone git@github.com:oliverbaehler/capsule.git
cd capsule & checkout issue/773

make docker-build
...

docker tag docker.io/clastix/capsule:v0.3.1 your.registry.com/capsule:dev
docker push your.registry.com/capsule:dev
```

Then update the capsule controller to use that image and remove ur temporary clusterRole. Let me know if that works or if you need further assistance.

**Comment by erikgb** [comment]
It's unfortunately a bit of an extra job to build and push an image to our internal registry. Are you able to push the image to your GHCR, or is it maybe already available from there?

**Comment by maxgio92** [comment]
I think the permission shouldn't be granted as the tenant owner shouldn't be allowed to operate on its `Tenant` resource.
TBH I still don't have a strong opinion on how to solve this.

**Comment by prometherion** [comment]
> It does not appear correct to grant a standard user access to update finalizers on a cluster-scoped resource.

Couldn't we solve that with the `resourceNames` key? It grants the required permissions only to those named resources.

**Comment by prometherion** [comment]
> Why must Capsule set Tenant as the Namespace **controller**? Would it be sufficient to set it as an owner?

There's a subtle but essential difference between OwnerReference and ControllerReference.

OwnerReference is used for garbage collection, thus the cascading deletion of resources.

ControllerReference adds to the OwnerReference the events propagation of owned resources to the owner.
tl;dr; when a Namespace is changed, the Tenant reconciliation is triggered, ensuring the desired state throughout the Namespaces of the Tenant.

> If we want to clean up (delete) namespaces owned by a tenant when the tenant is deleted, we would have to add a finalizer to tenants. Finalizers are a bit painful, but this could also add more flexibility in whether or not namespaces should be orphaned or not when a tenant is deleted. Related #563

I had a similar discussion with a guy at KubeCon EU 22 in Valencia, something as: __in a year, cascading deletion will be gone from Capsule__, it seems we're in late of a couple of months 🙃 

Jokes apart, I would discuss this topic in a different issue to keep simple the resolution of the main root cause, such as overcoming the `OwnerReferencesPermissionEnforcement` admission controller.

**Comment by prometherion** [comment]
Removing this issue from the v0.3.4 milestone since there's no support offered by the reporter and we (as maintainers) don't have the chance to test this functionally with a real OpenShift cluster.

To whoever is interested in this feature or support, please, reach out here to coordinate and lead the development, as well as the tests.

**Comment by erikgb** [comment]
> don't have the chance to test this functionally with a real OpenShift cluster.

I think there is a slight misunderstanding here. I am not asking for full Openshift support, and I understand that open-source projects like Capsule cannot (or will not) test the software on enterprise clusters like Openshift.

This issue is to support any Kubernetes cluster with the Kubernetes [OwnerReferencesPermissionEnforcement](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#ownerreferencespermissionenforcement) admission controller plugin enabled. This should be possible on any Kubernetes cluster, and I think it could be justified to enable this on CI e2e-tests in this project.

Openshift (and maybe other Kubernetes distros) enables this admission controller by default. That is the rationale behind this request.

**Comment by maxgio92** [comment]
Hi @erikgb @prometherion, sorry for my late response.
I'd keep this issue to track the implementation of the support for the [`OwnerReferencesPermissionEnforcement`](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#ownerreferencespermissionenforcement) admission controller, beyond Openshift.

Unfortunately, I didn't have much time to dig into it, so, I don't have useful details about that. But if you agree, I'd keep this issue open as the feature would provide benefit IMO.

**Comment by erikgb** [comment]
Thanks @maxgio92. We have put our adoption of Capsule on hold because of this. I would also appreciate this not being presented as "an Openshift problem". Capsule does not currently support an optional Kubernetes admission controller. 😉
